util/pingpong.c

```diff
@@ -0,0 +1,1940 @@
+			ct->remote_fi_addr, ctx);
+	return 0;
+}
+
+ssize_t ft_tx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size)
+{
+	ssize_t ret;
+
+	if (ft_check_opts(ct, FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE))
+		ft_fill_buf((char *) ct->tx_buf, size);
+
+	ret = ft_post_tx(ct, ep, size, &(ct->tx_ctx));
+	if (ret)
+		return ret;
+
+	ret = ft_get_tx_comp(ct, ct->tx_seq);
+	return ret;
+}
+
+ssize_t ft_post_inject(struct ct_pingpong *ct, struct fid_ep *ep, size_t size)
+{
+	FT_POST(fi_inject, ft_get_tx_comp, ct->tx_seq, "inject",
+			ep, ct->tx_buf, size,
+			ct->remote_fi_addr);
+	ct->tx_cq_cntr++;
+	return 0;
+}
+
+ssize_t ft_inject(struct ct_pingpong *ct, struct fid_ep *ep, size_t size)
+{
+	ssize_t ret;
+
+	if (ft_check_opts(ct, FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE))
+		ft_fill_buf((char *) ct->tx_buf, size);
+
+	ret = ft_post_inject(ct, ep, size);
+	if (ret)
+		return ret;
+
+	return ret;
+}
+
+
+ssize_t ft_post_rx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size, struct fi_context* ctx)
+{
+	FT_POST(fi_recv, ft_get_rx_comp, ct->rx_seq, "receive", ep, ct->rx_buf,
+			MAX(size, FT_MAX_CTRL_MSG),
+			fi_mr_desc(ct->mr),	0, ctx);
+	return 0;
+}
+
+ssize_t ft_rx(struct ct_pingpong *ct, struct fid_ep *ep, size_t size)
+{
+	ssize_t ret;
+
+	ret = ft_get_rx_comp(ct, ct->rx_seq);
+	if (ret)
+		return ret;
+
+	if (ft_check_opts(ct, FT_OPT_VERIFY_DATA | FT_OPT_ACTIVE)) {
+		ret = ft_check_buf((char *) ct->rx_buf, size);
+		if (ret)
+			return ret;
+	}
+	/* TODO: verify CQ data, if available */
+
+	/* Ignore the size arg. Post a buffer large enough to handle all message
+	 * sizes. ft_sync() makes use of ft_rx() and gets called in tests just before
+	 * message size is updated. The recvs posted are always for the next incoming
+	 * message */
+	ret = ft_post_rx(ct, ct->ep, ct->rx_size, &(ct->rx_ctx));
+	return ret;
+}
+
+/*
+ * fi_cq_err_entry can be cast to any CQ entry format.
+ */
+static int ft_spin_for_comp(struct fid_cq *cq, uint64_t *cur,
+			    uint64_t total, int timeout)
+{
+	struct fi_cq_err_entry comp;
+	struct timespec a, b;
+	int ret;
+
+	if (timeout >= 0)
+		clock_gettime(CLOCK_MONOTONIC, &a);
+
+	while (total - *cur > 0) {
+		ret = fi_cq_read(cq, &comp, 1);
+		if (ret > 0) {
+			if (timeout >= 0)
+				clock_gettime(CLOCK_MONOTONIC, &a);
+
+			(*cur)++;
+		} else if (ret < 0 && ret != -FI_EAGAIN) {
+			return ret;
+		} else if (timeout >= 0) {
+			clock_gettime(CLOCK_MONOTONIC, &b);
+			if ((b.tv_sec - a.tv_sec) > timeout) {
+				fprintf(stderr, "%ds timeout expired\n", timeout);
+				return -FI_ENODATA;
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int ft_get_cq_comp(struct fid_cq *cq, uint64_t *cur,
+			  uint64_t total, int timeout)
+{
+	int ret = ft_spin_for_comp(cq, cur, total, timeout);
+
+	if (ret) {
+		if (ret == -FI_EAVAIL) {
+			ret = ft_cq_readerr(cq);
+			(*cur)++;
+		} else {
+			FT_PRINTERR("ft_get_cq_comp", ret);
+		}
+	}
+	return ret;
+}
+
+int ft_get_rx_comp(struct ct_pingpong *ct, uint64_t total)
```

Comment:
I don't think we're using libfabric counters in this code.  Can probably be removed.